### ISSUE TITLE
http client and http.Transport IdleConnTimeout is not reliably doing DNS resolve

### ISSUE THREAD

**Issue by szuecs** [golang/go#23427]
### What version of Go are you using (`go version`)?

go version go1.9 linux/amd64

and 

go version go1.9.2 linux/amd64

### Does this issue reproduce with the latest release?

Yes

### What operating system and processor architecture are you using (`go env`)?

% go env                                                             
GOARCH="amd64"
GOBIN="/home/sszuecs/go/bin"
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/sszuecs/go"
GORACE=""
GOROOT="/usr/share/go"
GOTOOLDIR="/usr/share/go/pkg/tool/linux_amd64"
GCCGO="gccgo"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build505089582=/tmp/go-build -gno-record-gcc-switches"
CXX="g++"
CGO_ENABLED="1"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"

### What did you do?

I am running `go run main.go` and change the /etc/hosts to change www.google.de to 127.0.0.1.
Depending on  the order it fails to switch DNS after IdleConnTimeout or it will never request DNS again. DNS will be tried to lookup in case you first target it to 127.0.0.1 and after that comment the entry in /etc/hosts. The problem is that if you want to change your target loadbalancers via DNS lookup, this will not be done. The workaround is commented in the code, which reliably will do the DNS failover. The problem seems to be that IdleConnTimeout is bigger then the time.Sleep duration in the code, which you can also change to see that this works. In case of being an edge proxy with high number of requests, the case IdleConnTimeout < process-next-request will never happen.

    package main
    
    import (
    	"log"
    	"net"
    	"net/http"
    	"time"
    )
    
    func main() {
    
    	tr := &http.Transport{
    		DialContext: (&net.Dialer{
    			Timeout:   5 * time.Second,
    			KeepAlive: 30 * time.Second,
    			DualStack: true,
    		}).DialContext,
    		TLSHandshakeTimeout: 5 * time.Second,
    		IdleConnTimeout:     5 * time.Second,
    	}
    
    	go func(rt http.RoundTripper) {
    		for {
    			time.Sleep(1 * time.Second)
    
    			req, err := http.NewRequest("GET", "https://www.google.de/", nil)
    			if err != nil {
    				log.Printf("Failed to do request: %v", err)
    				continue
    			}
    
    			resp, err := rt.RoundTrip(req)
    			if err != nil {
    				log.Printf("Failed to do roundtrip: %v", err)
    				continue
    			}
    			//io.Copy(ioutil.Discard, resp.Body)
    			resp.Body.Close()
    			log.Printf("resp status: %v", resp.Status)
    		}
    	}(tr)
    
    	// FIX:
    	// go func(transport *http.Transport) {
    	// 	for {
    	// 		time.Sleep(3 * time.Second)
    	// 		transport.CloseIdleConnections()
    	// 	}
    	// }(tr)
    
    	ch := make(chan struct{})
    	<-ch
    }

### What did you expect to see?

I want to to see that IdleConnTimeout will reliably close idle connections, such that DNS is queried for new connections again, similar to the goroutine case in the code. We need to slowly being able to fade traffic.

### What did you see instead?

In case you start the application with /etc/hosts entry is not set, and then change it, it will never fail the request, so the new DNS lookup is not being made.

**Comment by meirf** [comment]
**Possible Signals**

If `persistConn.conn` is dialed via hostname:
- _TTL expiration_. If TTL expires, take action. This would require an API change since [`net`'s TTL doesn't appear to be exposed](https://github.com/golang/go/blob/ed6c6c9c11496ed8e458f6e0731103126ce60223/src/net/dnsmsg.go#L130).
- _Poll `LookupIP`_. Store initial ip set and keep updating with new polling response; if ip set changes (or more strictly if dialed ip is not longer in ip set), take action. How frequently would we poll? Are we willing to store this extra state, spend this time polling, using system resourced even with [optimizations](https://github.com/golang/go/issues/23866)?

**Action**

- _Close/mark for closure_. If the connection is idle, close it. If the connection is in use, mark it as not reusable. This would be similar to current usage of `persistConn.broken` though "broken" maybe too harsh language for this case. (The OP uses `CloseIdleConnections`, which "does not interrupt any connections currently in use", so calling `CloseIdleConnections` once could leave many connections on the old ips.)

I guess the use of "?" in the title was due to implementation complexity/propriety of addressing this in the standard library. Please let me know if I've gotten something wrong or if there are other issues I'm missing.

**Comment by szuecs** [comment]
@meirf For us `CloseIdleConnections` works pretty well, we use it in our internet facing central ingress http proxy and it's fast enough to do the DNS lookup while we do DNS based traffic switching (AWS route53 Weighted DNS records). If you would use `IdleConnTimeout` to call `CloseIdleConnections` would not do a breaking change of the API. I bet it would be in 99.999% of the cases fine, because  applications, also proxies, are most of the time CPU bound.
My question would be why you want to expose additional data if you already have `IdleConnTimeout`?

To answer your question about "?" in the title: I don't remember the why. I deleted it from the title. I am only digging from time to time into the Go src code, mostly in the `net` package. I guess with the "?" I wanted to reflect that. I hope it's now more clear. :)

**Comment by meirf** [comment]
tl;dr (1) Idle connections' awareness of dns changes appears to be absent from the API. (2) The easy part is forcing a redial. (3) IMO the hard part is what signal the code should use to redial.

My first comment was directed to bradfitz, who set the title to ["net/http: make Transport's idle connection management aware of DNS changes?"](https://github.com/golang/go/issues/23427#event-1421975849).

Based on my reading of your most recent comment, you think idle connections' awareness of dns changes is already promised by the current API:

> If you would use IdleConnTimeout to call CloseIdleConnections would not do a breaking change of the API.

I (and bradfitz based on his title) do not agree that idle connections' awareness of dns changes is already promised.

I think we agree that currently the only way to be aware of a dns change is to force a redial by not having connections kept alive, but IdleConnTimeout won't force a keep-alive connection to be closed unless your situation meet its definition. If you have a keep-alive connection and it never becomes idle, by definition it will never close itself due to IdleConnTimeout. _The cumulative amount of time a keep-alive connection is idle doesn't matter. The only way the keep-alive connection will be closed due to IdleConnTimeout is if it's idle for IdleConnTimeout consecutive/contiguous time._

1. To prevent any connections from being kept alive you can set DisableKeepAlives.
2. To increase the likelihood of a keep-alive connection being closed, you can decrease IdleConnTimeout.
3. To close all idle keep-alive connections, you can call CloseIdleConnections.

It's clear that all these would be helpful to you in redialing and therefore connecting to the new ip. This indirectly results in dns awareness (and possibly a lot of inefficiency). But in terms of the standard library, there is no promise of direct dns awareness.

**Comment by szuecs** [comment]
@meirf Ok I was not sure if questions were asked to me. As far as I understand now you did not asked me. :)
I put again the "?" in the title, because the change was not done by me (I should read the events). 
Thanks for your comments!

**Comment by dragonsinth** [comment]
You guys willing to take any patches on this issue?  This turns out to be a shockingly cross cutting problem for which workarounds are not great or perfect, and the default http transport basically guides you directly into this problem.

It doesn't seem that nutty to periodically re-resolve DNS for requested hostnames, and refuse to select an open conn whose IP is not in the current set for that hostname.

**Comment by rohitkeshwani07** [comment]
>Close/mark for closure. If the connection is idle, close it. If the connection is in use, mark it as not reusable. This would be similar to the current usage of persistConn.broken though "broken" may be too harsh language for this case. (The OP uses CloseIdleConnections, which "does not interrupt any connections currently in use", so calling CloseIdleConnections once could leave many connections on the old IPs.)

@meirf I think just implementing something like 
```
    // Keep-alive connections are closed after this duration.
    //
    // By default connection duration is unlimited.
    MaxConnDuration time.Duration
```
will guarantee that my connections are not used beyond a minute. So my servers can assume that DNS propagation takes 1 minute over and above TTL time.

This would be easy to implement.

**Comment by cnmcavoy** [comment]
We just hit this bug recently, and its very problematic in the kubernetes ecosystem. Kubernetes manages DNS via CoreDNS and updates records of services based on the readiness of the underlying pods in a cluster. However, with this bug, workloads that do not recheck DNS and re-use an idle connection make requests to unhealthy pods and then use garbage data from these connections and cause issues in the cluster. 

I was surprised that `http.Get` did not have a one to one relationship with DNS requests in Go applications, given that Go does not do DNS caching. The current `net.Transport` behavior is effectively caching DNS for applications that support idle connections, and worse, caching DNS indefinitely. I expected Go to recheck DNS for every http connection, even if it re-used an idle connection. I agree @meirf that using the TTL to reduce the number of DNS requests would be ideal, but if it's not easy to use the TTL, then checking every time would still be fine with me. 

This is very surprising behavior and not something that I can tune from the outside as a system administrator. My current workaround is to restart Go applications when DNS changes, but that is a brute-force solution and has many drawbacks.

**Comment by szuecs** [comment]
@cnmcavoy any http connection pool has some kind of dns caching besides maybe nodejs (sorry for bashing it). I think it's wrong to query dns more often than once in a while, because you want to have a connection open before you got http requests that trigger your calls. 
Feel free to change all your http.Client to use either the Transport replacement https://pkg.go.dev/github.com/zalando/skipper@v0.11.151/net#NewTransport
or the Client https://pkg.go.dev/github.com/zalando/skipper@v0.11.151/net#Client
Both are API compatible (best effort compatibility) drop in replacements with additional features you might want to have anyways in kubernetes.

**Comment by cnmcavoy** [comment]
>I think it's wrong to query dns more often than once in a while

@szuecs Why wouldn't the correct behavior here to be to respect the DNS record's TTL?

I'm aware that you can work around the existing behavior, although I haven't looked at the project you linked. However, that is not helpful for projects that are already built/packaged for the community on dockerhub. Repackaging OSS Golang applications for them to respect DNS in idle connections sounds like an odd thing to tell a user.

I have witnessed real-world cases of the idle connections being reused in a Go application _days_ after the DNS record was changed and out of date. That was very surprising to me and it seems like a bug in Golang.

**Comment by szuecs** [comment]
@cnmcavoy I am fully with you if you want to have a ttl based or "once in a while" based DNS Update to refresh newly created connections, but as issue creator I am against resolve every time as you also suggested.

**Comment by dragonsinth** [comment]
Periodic / TTL based would be wonderful.  Ideally, we would track the resolved ip for each connection in the idle pool, and when DNS results change, purge any connections whose ips are no longer in the list.  (Since a big use case involves IP sets that change incrementally as servers get rolled.)

**Comment by networkimprov** [comment]
cc @odeke-em @fraenkel since this is marked NeedsFix.

**Comment by szuecs** [comment]
@dragonsinth but in case of an ALB/NLB spanning 3 AZs (has 1 ip per AZ), you would constantly refresh, which is not a great idea.

**Comment by dragonsinth** [comment]
> @dragonsinth but in case of an ALB/NLB spanning 3 AZs (has 1 ip per AZ), you would constantly refresh, which is not a great idea. So purge cache as soon you see new ip is not as simple as it sounds. It can be done correctly but the change should make sure this case will be done correctly.

@szuecs would you mind de-acronym this for me?  I'm not a network engineer. :)

**Comment by marcind** [comment]
@dragonsinth when running an application fronted by an AWS Application or Network Load Balancer (ALB/NLB), the load-balancing layer might consist of more servers with distinct IP addresses than the number of records returned in a typical DNS query (this might be a function of how many Availability Zones the LB is spanned across or by how scaled out the LB servers are due to traffic). If that is the case, the set of IPs presented to a client doing frequent DNS queries will appear to change with every query, even though nothing in either the LB or the application that it is fronting has changed.

**Comment by dragonsinth** [comment]
Thanks!  Makes sense.  Maybe TTL can be used to help here?  So we purge connections whose IPs have not been resolved from DNS within their specified TTL?  This might actually improve global load balancing without too much churn.

**Comment by szuecs** [comment]
@dragonsinth Sorry, but I wonder if I miss read your comment https://github.com/golang/go/issues/23427#issuecomment-695226120, because you wrote about IP sets so more than one IP per idle connection, which sounds correct to me. I somehow understood the idea was to close the connection in case of an IP change, which would happen all the time if you have a hostname resolving in multiple IPs.

In practice it would be fine to purge only the idle connections that are connected to IPs, which are not being resolved anymore. We run since years a proxy with up to 500k rps that just closes regularly (every 30s) idle connections, so I guess it's fine to only do it for idle connections. It was always fast enough to move traffic fast to new endpoints.

And sorry for using AWS specific acronyms. Seems to not

**Comment by dragonsinth** [comment]
No worries.  Perhaps for even my own clarity I should spell it out a bit.  Imagine querying the same hostname repeatedly and getting different rotating answers:

```
0s: host.name { ip1, ip2, ip3 }, TTL 300
30s: host.name { ip2, ip3, ip4 }, TTL 300
60s: host.name { ip3, ip4, ip5 }, TTL 300
90s: host.name { ip4, ip5, ip1 }, TTL 300
120s: host.name { ip5, ip1, ip2 }, TTL 300
```

Basically we could just keep map of last seen ips for a hostname with resolved TTL.  After 120s it would look something like this:

```
"host.name":  map[ip]int {
  "ip1": <300s from now>,
  "ip2": <300s from now>,
  "ip3": <240s from now>,
  "ip4": <270s from now>,
  "ip5": <300s from now>,
}
```

When we go to borrow an idle connection from the pool, we throw it away if its associated IP has expired (or been removed from the map).

**Comment by szuecs** [comment]
@dragonsinth makes absolutely sense to me (type could be time.Duration instead of int, but the idea is great)!

**Comment by dragonsinth** [comment]
Or `expiry time.Time` -- you'd want to resolve an absolute timestamp when you get the DNS response so you can interpret it later.

**Comment by rohitkeshwani07** [comment]
> In practice it would be fine to purge only the idle connections that are connected to IPs, which are not being resolved anymore.
@szuecs 
[https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/routing-policy.html](weighted dns) will return different IPs  but that doesn't mean old IPs are not valid anymore.

**Comment by szuecs** [comment]
@rohitkeshwani07 same is if you want to rotate IPs for whatever reason, old ones can still be valid. Maybe infrastructure just observes until no traffic and then stops. So maybe we should resolve again to switch traffic faster, but I don't know. It's probably fine to just use DNS TTL expiry.

**Comment by rohitkeshwani07** [comment]
To what I understand, there are 3 possible solutions 

1. In case IP set changes, purge all connections and those in use should be marked not reusable (change in IP set will be done periodically every X seconds)
Con: if I am using weightedDNS, then probably my connections will be rotated every X seconds which is performance overhead for backends as they will have to do SSL handshake again for new connections.
2. TTL based expiry
Con: if for some reason, I have set high TTL in my DNS. Connections won't be rotated ever, which again might raise problems in a scenario where your client assumes that the connection is open but the server has actually closed the connection. Whenever we try to use that half-open connection, we get socket timeout.
3. max connection lifetime (i vote for)
This will assume that a connection once created, will be used for a limited lifetime, and hence new DNS IPs will be resolved after this time. 

Also, we set max_connection_lifetime equal to max_connections, It can be assumed that there will be close to 1 connection establishment per second which will be good for performance-intensive applications.

This is also the way other JAVA HTTP clients are solving the problem. 

// @szuecs

**Comment by szuecs** [comment]
@rohitkeshwani07 3. seems also fine for me. I would not do 1. If popular JAVA clients do 3., I think we can safely assume it's not a bad choice and the behavior is well understood. To not get socket timeouts you can always work with backends together to get an aligned time.Duration (client 1s lower than the backend and you won't get any timeouts for this case).

**Comment by rv-rhill** [comment]
.NET also offers [the exact some configuration option](https://docs.microsoft.com/en-us/dotnet/api/system.net.servicepoint.connectionleasetimeout?view=netcore-3.1#System_Net_ServicePoint_ConnectionLeaseTimeout) described in @rohitkeshwani07's solution 3. 

My organization faces the same problem originally articulated by @szuecs, since we're using a Route 53 Latency Policy record to handle dynamic, latency-optimized routing to ELB's + ECS services running the same app in multiple regions. I think this opportunity to differentiate between timeouts for a) connections that have fallen into idle state and b) connections in active (re)use will make it much easier to solve for the common problem of detecting changes in the IP set for a single DNS record over time. 

It will also more clearly signal, in the docs for package `http`, that not all _reusable_ connections necessarily become _idle_. I've noticed from some experimentation with the fabulous `httptrace` package that Go's `http.Transport` can efficiently reuse a greater number of connections per host than the specified `MaxIdleConnsPerHost`. If I run a group of concurrent goroutines all using the same underlying `http.Transport` to send requests continually to the same host, I see considerably more connections re-used across those goroutines than the `MaxIdleConns`. That is to say, many more of the connections reported in the `GotConn` [hook](https://golang.org/pkg/net/http/httptrace/#ClientTrace) were reused and not idle than reused and idle. Still trying to track down in the source code whether it's true, as it seems, that a persistent connection can be reused for subsequent requests before being put to the idle connection pool.

**Comment by cnmcavoy** [comment]
@rohitkeshwani07 3. sounds like it would fit my needs as well.

Back when I first encountered this issue, [I wrote up a Golang proof-of-concept to reproduce](https://github.com/indeedeng-alpha/Golang-DNS-resolution-bug-proof-of-concept), and I had it open sourced. It runs a pure Go DNS server and makes requests on two interfaces to demonstrate the incorrect behavior. It is hopefully simpler to use to reproduce Golang's behavior than the steps in the OP & /etc/hosts mangling. The README.md has more details.

**Comment by krishnakowshik** [comment]
@cnmcavoy @rohitkeshwani07 I am facing this issue as well. How do you set the http max connection lifetime ? I am setting IdleConnTimeout and I am setting transport config to - 

Transport: &http.Transport{
			DialContext: (&net.Dialer{
				Timeout:   20 * time.Second,
				KeepAlive: 20 * time.Second,
			}).DialContext,
			IdleConnTimeout: 2 * time.Second,			
		}

**Comment by szuecs** [comment]
@krishnakowshik right now you have to close the idle connections yourself as I wrote in the issue.

**Comment by kingcr7** [comment]
> @krishnakowshik right now you have to close the idle connections yourself as I wrote in the issue.
I think it would not happen when disable KeepAlive .Only set enable KeepAlive ,it would happen。
Should i set a job call CloseIdleConnections Once per minute? or when http response error then call CloseIdleConnections?

**Comment by RobGraham** [comment]
We recently upgraded from **1.15.5** to **1.15.7**. We started noticing the same connection issues experienced PRIOR to the introduction of the fix defined in this thread... We ultimately had to rollback and since then haven't experienced any problems. 
Reviewing the release notes of 15.6 and 15.7, i saw no obvious net related changes. Has anyone else experienced issues with these versions?

**Comment by josephcopenhaver** [comment]
I've used MaxConnectionLifespan in other transports to ensure dns resolution and load-balancing got refreshed consistently. Took a stab at it here because it looks like the easiest way to implement it is to submit a PR to the standard lib.

https://github.com/golang/go/pull/46714

**Comment by jpcope** [comment]
Rather than hardcoding the lifespan a boolean could be placed in the transport that tells the connection to inherit the reuseDeadline from a dns response TTL.

Obviously that would need to be piped up somehow from the dialer. Since it passes the context around it could be leveraged to pipe back a response thus everything remains backwards compatible.

**Comment by neild** [comment]
https://golang.org/cl/341329 has a possible approach that @bradfitz and I came up with.

The idea is to add a `ReuseConn` hook to `http.Transport` that permits user-specified policy decisions on when to reuse a connection from an idle pool. Setting a max lifetime would be something like:

```
transport.ReuseConn = func(info httptrace.ReuseConnInfo) bool {
  return time.Since(info.CreationTime) < MaxConnLifetime
}
```

**Comment by jpcope** [comment]
I like the simplicity of the extra hook. I think I prefer the deadline approach which is used in the idle timeout purely because I know the client is not going to keep the connection around. So don't need a usage attempt to prompt the connection to close.

Would you need to add similar functionality in http2's pool?

**Comment by szuecs** [comment]
@neild for me personally `transport.ReuseConn` is great and `ClientTrace.ReuseConn` is not really required, because you would already see in the ClientTrace, that a new conn is created (DNSStart, DNSDone, ConnectStart, ConnectDone and potentially also the TLS parts). It feels also a bit weird to me that clienttrace can now influence the connection pool, which seems more the responsibility of the `http.Transport`, but maybe it's just my first thought and it is just fine.

**Comment by jpcope** [comment]
@neild One of the reasons I am all for adding a MaxConnLifespan to the Transport and socket wait logic is that a given http service which brokers connections to other services can start truncating connections without requiring a new connection/request to be made to a service. This in theory would free up connections for any backing services as horizontal scale-out ( runtime replication ) events occur and potentially reduce the number of idle connections at any given time across the gamut of services. We've implemented this logic before in other protocols and saw clear performance gains as well as the dns re-resolution benefits.

Please consider the PR I have out there as well. It's my first one so I don't mind it ultimately being closed because it has style issues. Would be ecstatic to see the capability in the standard lib even if my name is not attached. 😄 

Gonna resolve the conflict in the api/next.txt file again in a sec.

**Comment by aojea** [comment]
> https://golang.org/cl/341329 has a possible approach that @bradfitz and I came up with.
> 
> The idea is to add a `ReuseConn` hook to `http.Transport` that permits user-specified policy decisions on when to reuse a connection from an idle pool. Setting a max lifetime would be something like:
> 
> ```
> transport.ReuseConn = func(info httptrace.ReuseConnInfo) bool {
>   return time.Since(info.CreationTime) < MaxConnLifetime
> }
> ```

@neild assuming http2 is the most impacted by this issue since it will multiplex the requests over a single connection, not dialing again, and that the ClientConnPool already is an interface

https://github.com/golang/net/blob/ef0fda0de5086a0a3c6721b466f49e1f36188234/http2/client_conn_pool.go#L17-L27

what if you can expose the clientConnPool type so people can implement their own logic on  `GetClientConn()`

```go

type myClientConnPool {
net.ConnectionPool
}

func (p *myClientConnPool) GetClientConn(req *http.Request, addr string) (*ClientConn, error) {
        ////
	return p.GetClientConn(req,addr)
}

func (p *myClientConnPool) MarkDead(c *ClientConn) {
	p.MarkDead(c)
}
```

I think that the the connection pool type can benefit other issues like https://github.com/golang/go/issues/37373 , this last issue may require some additional changes since the conn poll is keyed by `host:port` but just wanted to know your thoughts

**Comment by chengzhicn** [comment]
>     2. TTL based expiry
>        Con: if for some reason, I have set high TTL in my DNS. Connections won't be rotated ever, which again might raise problems in a scenario where your client assumes that the connection is open but the server has actually closed the connection. Whenever we try to use that half-open connection, we get socket timeout.

@rohitkeshwani07 In case of high TTL, redial probably won't work because of DNS cache in host/resolver. If the main concern is DNS change, then it's better to use TTL because:
1. It's a standard way to control how long the information should stay valid.
2. Both TTL and DNS change is done by the domain owner, and he can decide how long the TTL will be since different value will have different impact on authoritative DNS server(more/less query because of cache expires) and application server(more/less redial and handshake). He can set a large TTL if he want less load on his servers and accept the fact that DNS change will take a long time to take effect.

But if the main problem resides on client side(eg: firewall drops longstanding connections), then it make sense to take client side approach like configure max connection lifetime or hooks proposed by @neild on client side.

**Comment by josephcopenhaver** [comment]
Hello again!

I still really like adding logic to the transport layers so system resources can close in advance of trying to pull a connection from the pool should a deadline be reached: https://github.com/golang/go/pull/46714

This behavior is more ideal for me just because then I don't have system nodes keeping connections alive just so the pool layer on top can intercept an established connection candidate, see it is too old, and then kill it. May as well let it take care of itself in that regard and short circuit earlier.

Any way I can help ensure an approach of some kind makes it into the standard lib? 😄 I think it would be very powerful to get the standard http kit up to par with or allow for the creation of the same knobs we have in https://pkg.go.dev/database/sql

**Comment by josephcopenhaver** [comment]
I wonder how much of this problem is purely one of loadbalancing concepts people may be trying to implement.

If a host record resolves to 3 ip nodes and suddenly we add or remove one, then respectively we need to ensure the new node starts taking requests as soon as we know about it so load spreads ( no need to close connections until we approach the max connection limit or idle timeouts happen ) or remove connections tied to the deregistering ( potentially already deregistered ) ip address.

The latter case of removing a node is trivial and fairly self-correcting before any DNS TTL and refresh cycle would inform us on average.

In the same token, idle connection timeout can create "randomly hot" nodes over time by just the probability of FIFO rotating through connections cyclicly. A max lifespan is a fairly nuclear solution to this problem when perhaps envoy proxy should be the full-featured solution for balancing out-of-app.

For people wanting a solution in the same binary they really probably want LIFO connection queuing under a round-robin ( or some distance/efficiency-aware weighted function ) connection broker over the LIFO queues partitioned by IP. This would allow for any "logically extra" connections the service can live without to end up dying off just by virtue of the connection timeouts.

If DNS appears to remove a host we can assume the DNS source has determined that the host should get phased out and schedule the relevant connections to die as soon as their transactions finish or immediately inform the broker to kill that pool of trusted connections.

As a DNS entry is discovered for a hostname a LIFO queue would be created and added to the connection broker. Load would spread almost immediately across the upstream offering, and old connections no longer getting any use on the other nodes would start to time out and die off, expensive new connection flows continue to be avoided.

Connection max lifetime/reuse is not the feature people need or want. It's just what we're reaching for in this conversation when we really want to observe+schedule DNS resolution and use that to inform a connection pooling strategy where we can change the implementation of popping and pushing to be LIFO or FIFO or anything else. It's worth discussing in more detail by persons probably smarter than myself. Typically users of one of my libs want agency I will happily give them the simplest interface they need to achieve their goals.

**Comment by szuecs** [comment]
Thanks @josephcopenhaver for your answer. A lot of things make sense to me. The problem is not proxy specific, more http upstream specific.
If you have an url as target and the target is not an anycast ip, then you have some number of IPs and if the target needs to scale out it needs more IPs, so if clients stay the new ones won't be used. The lifo idea might be fixing this problem, so maybe worth a try.

**Comment by marcind** [comment]
> I wonder how much of this problem is purely one of loadbalancing concepts people may be trying to implement.

Load balancing is not the reason why I'm interested in this issue. My use case: a long-running workload that has a steady cadence of requests to a HTTP 1.1 service `Foo` (a few/second, maybe it's logs getting uploaded, etc). `Foo`'s operator has configured a low DNS TTL to ensure traffic can be moved quickly if needed. At the same time, the workload uses HTTP keep-alive to avoid redialing for each request. `Foo`'s operator updates their DNS in hopes of inducing a graceful blue/green move to new infrastructure and yet observes little to no traffic shifting in response, because the steady request cadence ensures no connections go idle. Ideally, the old `Foo` endpoints would start responding with `Connection: close` headers, but... `Foo` is fronted by an AWS Application Load Balancer and there exists no mechanism to induce an ALB to start closing client connections.

> If DNS appears to remove a host we can assume the DNS source has determined that the host should get phased out and schedule the relevant connections to die as soon as their transactions finish or immediately inform the broker to kill that pool of trusted connections.

That is not a safe assumption. For example, DNS responses sent by an AWS ALB will cycle through a subset of the existing IP addresses. Just because an IP does not appear in a subsequent DNS query does not mean that the corresponding instance no longer accepts traffic. In fact, that IP will most likely reappear in a following DNS response.

> Connection max lifetime/reuse is likely not the feature people need or want. ( If you do need/want it cool, my PR offers it! ) It just looks like what we're reaching for in this conversation when we really want to observe+schedule DNS resolution and use that to inform a connection pooling strategy where we can change the implementation of popping and pushing to a queue we maintain externally that is LIFO or FIFO or anything else - but note we need to know the connection's target IP to achieve these goals. It's worth discussing in more detail by persons probably smarter than myself. Typically when users of one of my libs want agency I will happily give them the simplest interface they need to achieve their goals and default the behavior to off. Perhaps the maintainers are willing to do similar here?

Yes and no. Years of "best practices" have conditioned everyone to expect that setting a low DNS TTL will allow service operators to induce HTTP clients to shift traffic with minimum delay. Many browsers and software SDKs behave that way (even though there's nothing in the HTTP specification that mandates keep-alives honor DNS TTLs), but `net/http` does not. So on the one hand I would be happy to have the HTTP client track the remaining TTL deadline that was associated with the DNS response used to establish a particular connection. On the other, I would still like the ability to extend that connection's life (assuming the server indicates that this is OK by not returning `Connection: close`) beyond the 30 seconds that's typically configured for these sort of applications. Setting a MaxLIfetime of say 5 minutes would strike a good balance between "immediately" (30 seconds) and "indefinitely" (i.e. whenever the service operator finally decides to just shut down the "old" deployment since the existing clients don't seem to want to shift gracefully).

**Comment by szuecs** [comment]
okhttp suggested to do this on the server side and send `Connection: close` response header regularly, maybe similar to my time based suggestion on the client: https://github.com/square/okhttp/issues/3374#issuecomment-307569702
In the end I think both (client and server side solutions) have their advantages and disadvantages.

**Comment by awnumar** [comment]
@szuecs while it is possible to solve this server-side, sometimes the server-side application is owned by a third party so it's not within our control to make changes to it.

Our solution for working around the lack of a client-side connection-lifetime timeout is:

- have a global http client
- have a routine that regularly initialises a new client and replaces the global one
- new requests use the new client which opens a new connection

**Comment by szuecs** [comment]
@awnumar this is also possible and I think it would be beneficial to have it on the client but I wanted also to show that there are others that show server side solutions in this context. A chosen solution depends on the context.

**Comment by andrebrait** [comment]
@szuecs just contributing here with my 2 cents: I recently attempted the suggested solution by the OkHttp authors, and it does work, but not with OkHttp because it doesn't actually seem to run name resolution again until it actually gets an error from the server (e.g. a 503). 

The server-side approach seems to still require some proper handling on the client to achieve awareness of DNS record changes.

**Comment by szuecs** [comment]
@andrebrait thanks! We also added https://github.com/zalando/skipper/pull/3246 to fix this similar to other proxies. okhttp seems to be unfortunate in handling this case. :(
